core/src/main/java/org/nervos/ckb/transaction/CellGatherer.java

```diff
@@ -0,0 +1,83 @@
+package org.nervos.ckb.transaction;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+import org.nervos.ckb.methods.type.Script;
+import org.nervos.ckb.methods.type.cell.CellInput;
+import org.nervos.ckb.methods.type.cell.CellOutputWithOutPoint;
+import org.nervos.ckb.service.CKBService;
+import org.nervos.ckb.system.type.SystemScriptCell;
+
+/** Copyright © 2019 Nervos Foundation. All rights reserved. */
+public class CellGatherer {
+
+  private CKBService ckbService;
+
+  public CellGatherer(CKBService ckbService) {
+    this.ckbService = ckbService;
+  }
+
+  Cells getCellInputs(String lockHash, BigInteger needCapacities) throws IOException {
+    List<CellInput> cellInputs = new ArrayList<>();
+    BigInteger inputsCapacities = BigInteger.ZERO;
+    long toBlockNumber = ckbService.getTipBlockNumber().send().getBlockNumber().longValue();
+    long fromBlockNumber = 1;
+
+    while (fromBlockNumber <= toBlockNumber && inputsCapacities.compareTo(needCapacities) < 0) {
+      long currentToBlockNumber = Math.min(fromBlockNumber + 100, toBlockNumber);
+      List<CellOutputWithOutPoint> cellOutputs =
+          ckbService
+              .getCellsByLockHash(
+                  lockHash, String.valueOf(fromBlockNumber), String.valueOf(currentToBlockNumber))
+              .send()
+              .getCells();
+
+      if (cellOutputs != null && cellOutputs.size() > 0) {
+        for (CellOutputWithOutPoint cellOutputWithOutPoint : cellOutputs) {
+          CellInput cellInput = new CellInput(cellOutputWithOutPoint.outPoint, "0");
+          inputsCapacities =
+              inputsCapacities.add(new BigDecimal(cellOutputWithOutPoint.capacity).toBigInteger());
+          cellInputs.add(cellInput);
+          if (inputsCapacities.compareTo(needCapacities) > 0) {
+            break;
+          }
+        }
+      }
+      fromBlockNumber = currentToBlockNumber + 1;
+    }
+    return new Cells(cellInputs, new BigDecimal(inputsCapacities).toBigInteger());
+  }
+
+  public BigInteger getCapacitiesWithPrivateKey(String privateKey) throws IOException {
```

Comment:
Using `privateKey` to getBalance seems strange.
I think using `address/lockHash/publicKey` is enough.

Reply:
Fixed https://github.com/nervosnetwork/ckb-sdk-java/pull/152/commits/3aee99614bdd9915dbebb2191090c0c494cb552c